openssl/src/bn.rs

```diff
@@ -868,13 +868,46 @@ impl BigNumRef {
     /// assert_eq!(BigNum::from_slice(&s_vec).unwrap(), r);
     /// ```
     pub fn to_vec(&self) -> Vec<u8> {
-        let size = self.num_bytes() as usize;
-        let mut v = Vec::with_capacity(size);
+        let mut v = vec![0u8; self.num_bytes() as usize];
+        self.to_slice(&mut v).unwrap();
+        v
+    }
+
+    /// Encodes the absolute value of `self` into a big-endian buffer.
+    ///
+    /// `self` can be recreated by using `from_slice`.
+    ///
+    /// This method fails if the buffer is not large enough. If the
+    /// buffer is larger than necessary, the number is prefixed by
+    /// zeroes so that the result is always a valid big-endian number.
```

Comment:
We generally try to bind OpenSSL functions "directly", so I think it makes more sense to write to the head of the buffer and return the number of bytes written like the underlying API does.

Reply:
This makes it hard to write into fixed array sizes, which is my motivation for writing this patch. For example, I need to write an RSA number into a 384 byte buffer to hand to hardware. I don't control the size of the buffer.

If we write to the beginning of the buffer, now I have to manually move it and zero-pad. This imposes the cost of an extra `memmove()` on every consumer, all to retain exact compatibility with a terrible OpenSSL interface that doesn't take a length parameter to a buffer it writes to (!).

I understand the desire to mimic the OpenSSL interface. But this particular interface is a pretty bad one.

Reply:
Why do you have to memmove? Just slice down to the tail of the buffer.

Reply:
Yes, but now I have to carry a utility function that checks the length of the number and splits the buffer - doing zeros in the first half and passing the second half to this function. This also implies unsafe code for performance. I'd rather consolidate this code.

Reply:
This seems like an extremely niche use case to your hardware thing.

Reply:
Fixed sized integer conversions are extremely common the moment you do crypto in hardware. There is a lot of crypto hardware out there.

Should we have two methods: one that wraps the OpenSSL behavior and one that writes a valid big endian number to the size of the buffer?

Reply:
We should expose the OpenSSL interfaces that exist. If OpenSSL has not seen it fit to expose the interface you want, it is not clear to me why this library should.

I would also be interested to see how you've quantified the performance costs that require the use of the unsafe write_bytes function rather than just writing zeroes in a loop. AFAICT LLVM will recognize that idiom and lower to a memset anyway.

Reply:
Okay. I will expose the OpenSSL interface. I will also wrap it in `impl TryFrom`. The only thing left to agree on is what error to use for `TryFrom` when the `BigNum` is too large for the target integer type.

Reply:
@sfackler It appears that OpenSSL *does* provide a padding variant: `BN_bn2binpad()`. I'll wrap this function too.